runtime/memprof.c

```diff
@@ -77,55 +76,108 @@ static intnat callstack_buffer_len = 0;
 
 /**** Statistical sampling ****/
 
-static double mt_generate_uniform(void)
-{
+Caml_inline uint32_t mt_shuffle(uint32_t a, uint32_t b, uint32_t c) {
+    uint32_t y = (a & 0x80000000) + (b & 0x7fffffff);
+    return c ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
+}
+
+/* This function regenerates [MT_STATE_SIZE] geometric random
+   variables at once. Doing this by batches help us gain performances:
+   many compilers (e.g., GCC, CLang, ICC) will be able to use SIMD
+   instructions to get a performance boost.
+*/
+#ifdef SUPPORTS_TREE_VECTORIZE
+__attribute__((optimize("tree-vectorize")))
+#endif
+static void rand_batch(void) {
   int i;
-  uint32_t y;
 
-  /* Mersenne twister PRNG */
-  if (mt_index == MT_STATE_SIZE) {
-    for (i = 0; i < 227; i++) {
-      y = (mt_state[i] & 0x80000000) + (mt_state[i+1] & 0x7fffffff);
-      mt_state[i] = mt_state[i+397] ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
-    }
-    for (i = 227; i < MT_STATE_SIZE - 1; i++) {
-      y = (mt_state[i] & 0x80000000) + (mt_state[i+1] & 0x7fffffff);
-      mt_state[i] = mt_state[i-227] ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
-    }
-    y = (mt_state[MT_STATE_SIZE - 1] & 0x80000000) + (mt_state[0] & 0x7fffffff);
-    mt_state[MT_STATE_SIZE - 1] =
-      mt_state[396] ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
-    mt_index = 0;
+  /* Instead of using temporary buffers, we could use one big loop,
+     but it turns out SIMD optimizations of compilers are more fragile
+     when using larger loops.  */
+  static uint32_t buff32[MT_STATE_SIZE];
+  static float bufff[MT_STATE_SIZE];
```

Comment:
I am not a fan of your variable  names. `bufff` is especially painful, being neither meaningful nor easy to read and type.  Do you have meaningful names?  Otherwise, `A` and `B` would work just as well.